hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/parser/record/transactionhandler/EthereumTransactionHandler.java

```diff
@@ -0,0 +1,155 @@
+import com.hedera.mirror.common.domain.transaction.TransactionType;
+import com.hedera.mirror.common.util.DomainUtils;
+import com.hedera.mirror.importer.domain.EntityIdService;
+import com.hedera.mirror.importer.parser.record.entity.EntityListener;
+import com.hedera.mirror.importer.parser.record.entity.EntityProperties;
+import com.hedera.mirror.importer.parser.record.ethereum.EthereumTransactionParser;
+
+@Named
+@RequiredArgsConstructor
+class EthereumTransactionHandler implements TransactionHandler {
+    private final EntityIdService entityIdService;
+    private final EntityProperties entityProperties;
+    private final EntityListener entityListener;
+    private final EthereumTransactionParser ethereumTransactionParser;
+
+    /**
+     * Attempts to extract the contract ID from the ethereumTransaction.
+     *
+     * @param recordItem to check
+     * @return The contract ID associated with this ethereum transaction call
+     */
+    @Override
+    public EntityId getEntity(RecordItem recordItem) {
+        var transactionRecord = recordItem.getRecord();
+
+        // pull entity from ContractResult
+        var contractFunctionResult = transactionRecord.hasContractCreateResult() ?
+                transactionRecord.getContractCreateResult() : transactionRecord.getContractCallResult();
+
+        EntityId contractId = null;
+        EntityId senderId = null;
+        EthereumTransaction ethereumTransaction = null;
+        if (!contractFunctionResult.equals(ContractFunctionResult.getDefaultInstance())) {
+            contractId = entityIdService.lookup(contractFunctionResult.getContractID());
+            senderId = EntityId.of(contractFunctionResult.getSenderId());
+
+            // no contractResult signifies a non executed transaction
+            ethereumTransaction = insertEthereumTransaction(recordItem, senderId);
+        }
+
+        if (transactionRecord.hasContractCreateResult()) {
+            insertContract(contractId, senderId, ethereumTransaction, recordItem.getConsensusTimestamp());
+        }
+
+        return contractId;
+    }
+
+    @Override
+    public TransactionType getType() {
+        return TransactionType.ETHEREUMTRANSACTION;
+    }
+
+    private EthereumTransaction insertEthereumTransaction(RecordItem recordItem, EntityId senderId) {
+        if (!entityProperties.getPersist().isEthereumTransactions()) {
+            return null;
+        }
+
+        var body = recordItem.getTransactionBody().getEthereumTransaction();
+        var ethereumDataBytes = DomainUtils.toBytes(body.getEthereumData());
+        var ethereumTransaction = ethereumTransactionParser.parse(ethereumDataBytes);
+        if (ethereumTransaction == null) {
+            return null;
+        }
+
+        // update ethereumTransaction with body values
+        var file = body.getCallData() == FileID.getDefaultInstance() ? null : EntityId.of(body.getCallData());
+        ethereumTransaction.setCallDataId(file);
```

Comment:
We should always use the `has` methods if available.

```suggestion
        if (body.hasCallData()) {
          ethereumTransaction.setCallDataId(EntityId.of(body.getCallData()));
        }
```

Reply:
Updated